lib/datadog/appsec/contrib/rails/gateway/watcher.rb

```diff
@@ -0,0 +1,81 @@
+# typed: false
+
+require 'datadog/appsec/instrumentation/gateway'
+require 'datadog/appsec/reactive/operation'
+require 'datadog/appsec/contrib/rails/reactive/action'
+require 'datadog/appsec/event'
+
+module Datadog
+  module AppSec
+    module Contrib
+      module Rails
+        module Gateway
+          # Watcher for Rails gateway events
+          module Watcher
+            def self.watch
+              Instrumentation.gateway.watch('rails.request.action') do |stack, request|
+                block = false
+                event = nil
+                waf_context = request.env['datadog.waf.context']
+
+                AppSec::Reactive::Operation.new('rails.request.action') do |op|
+                  trace = active_trace
+                  span = active_span
+
+                  Rails::Reactive::Action.subscribe(op, waf_context) do |action, result, _block|
+                    record = [:block, :monitor].include?(action)
+                    if record
+                      # TODO: should this hash be an Event instance instead?
```

Comment:
I've seen this so many times that I think the answer is probably "yes" now :)

Reply:
Damn yes.